### ISSUE TITLE
when using host_vector<glm::vec4>: warning: calling a __host__ function from a __host__ __device__ function is not allowed

### ISSUE THREAD

**Issue by cjue** [thrust/thrust#875]
Test case:  `/usr/local/cuda-8.0/bin/nvcc test_thrust_warning.cu -c`
~~~
#include <thrust/host_vector.h>
#include <glm/glm.hpp>

int main(void)
{
        thrust::host_vector<glm::vec4> colors;
        colors.push_back(glm::vec4(1.f, 1.f, 0.f, 1.f));
}
~~~

On Ubuntu Trusty with Cuda 7.5 this compiles without warning.

On Ubuntu Xenial with Cuda 8.0, Thrust 1.8.3 and GLM 0.9.7.2 I get this warning:
```
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/allocator/allocator_traits.inl(110): warning: calling a __host__ function from a __host__ __device__ function is not allowed
          detected during:
            instantiation of "thrust::detail::enable_if<thrust::detail::allocator_traits_detail::has_member_construct2<Alloc, T, Arg1>::value, void>::type thrust::detail::allocator_traits_detail::construct(Alloc &, T *, const Arg1 &) [with Alloc=std::allocator<glm::vec4>, T=glm::vec4, Arg1=glm::vec4]" 
(278): here
            instantiation of "void thrust::detail::allocator_traits<Alloc>::construct(thrust::detail::allocator_traits<Alloc>::allocator_type &, T *, const Arg1 &) [with Alloc=std::allocator<glm::vec4>, T=glm::vec4, Arg1=glm::vec4]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/allocator/copy_construct_range.inl(54): here
            instantiation of "void thrust::detail::allocator_traits_detail::copy_construct_with_allocator<Allocator, InputType, OutputType>::operator()(Tuple) [with Allocator=std::allocator<glm::vec4>, InputType=glm::vec4, OutputType=glm::vec4, Tuple=thrust::detail::tuple_of_iterator_references<glm::vec4 &, glm::vec4 &, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/function.h(60): here
            instantiation of "Result thrust::detail::wrapped_function<Function, Result>::operator()(const Argument &) const [with Function=thrust::detail::allocator_traits_detail::copy_construct_with_allocator<std::allocator<glm::vec4>, glm::vec4, glm::vec4>, Result=void, Argument=thrust::detail::tuple_of_iterator_references<glm::vec4 &, glm::vec4 &, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/system/detail/sequential/for_each.h(56): here
            instantiation of "InputIterator thrust::system::detail::sequential::for_each(thrust::system::detail::sequential::execution_policy<DerivedPolicy> &, InputIterator, InputIterator, UnaryFunction) [with DerivedPolicy=thrust::system::cpp::detail::tag, InputIterator=thrust::zip_iterator<thrust::tuple<thrust::detail::normal_iterator<glm::vec4 *>, glm::vec4 *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>, UnaryFunction=thrust::detail::allocator_traits_detail::copy_construct_with_allocator<std::allocator<glm::vec4>, glm::vec4, glm::vec4>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/for_each.inl(44): here
            [ 4 instantiation contexts not shown ]
            instantiation of "thrust::detail::contiguous_storage<T, Alloc>::iterator thrust::detail::contiguous_storage<T, Alloc>::uninitialized_copy(InputIterator, InputIterator, thrust::detail::contiguous_storage<T, Alloc>::iterator) [with T=glm::vec4, Alloc=std::allocator<glm::vec4>, InputIterator=thrust::detail::normal_iterator<glm::vec4 *>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/vector_base.inl(833): here
            instantiation of "void thrust::detail::vector_base<T, Alloc>::fill_insert(thrust::detail::vector_base<T, Alloc>::iterator, thrust::detail::vector_base<T, Alloc>::size_type, const T &) [with T=glm::vec4, Alloc=std::allocator<glm::vec4>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/vector_base.inl(593): here
            instantiation of "void thrust::detail::vector_base<T, Alloc>::insert(thrust::detail::vector_base<T, Alloc>::iterator, thrust::detail::vector_base<T, Alloc>::size_type, const T &) [with T=glm::vec4, Alloc=std::allocator<glm::vec4>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/vector_base.inl(581): here
            instantiation of "thrust::detail::vector_base<T, Alloc>::iterator thrust::detail::vector_base<T, Alloc>::insert(thrust::detail::vector_base<T, Alloc>::iterator, const T &) [with T=glm::vec4, Alloc=std::allocator<glm::vec4>]" 
/usr/local/cuda-8.0/bin/../targets/x86_64-linux/include/thrust/detail/vector_base.inl(495): here
            instantiation of "void thrust::detail::vector_base<T, Alloc>::push_back(const thrust::detail::vector_base<T, Alloc>::value_type &) [with T=glm::vec4, Alloc=std::allocator<glm::vec4>]" 
test_thrust_warning.cu(7): here
```
The only similar bug I could find was issue #741 (with pull request #743).

**Comment by elalish** [comment]
I'm getting the same thing, but with CUDA 8.0 on OSX. I get this warning the first time I use a thrust::host_vector function (.resize, .push_back, etc.). It'll only throw the warning once though. And `#pragma hd_warning_disable` does not remove it.

**Comment by dzhwinter** [comment]
any update of this issue?

**Comment by brycelelbach** [comment]
Hey, it's on my plate, but I probably won't be able to look at this until early October.

**Comment by brycelelbach** [comment]
Okay, looking into this.

**Comment by brycelelbach** [comment]
This is fixed in Thrust 1.9.0 - aka the version of Thrust in CUDA 9.0. Verified no warnings on CUDA 9.0.167/Debian 9.1/Linux 4.9.0-3 + r390 nvidia driver/i7-7820 + GM107.

The CUDA backend was rewritten for Thrust 1.9.0. Currently, 1.9.0 is not pushed to Github, but you can get it the [CUDA 9.0 toolkit](https://developer.nvidia.com/cuda-toolkit/whatsnew).

Closing as fixed, re-open if needed.
